src/onegov/fsi/upgrade.py

```diff
@@ -172,3 +174,125 @@ def add_active_property_to_attendees(context):
+
+    if m := p3.match(address):
+        postal_addr = m.group(1) + '\n' + m.group(2)
+        postal_pcc = m.group(3)
+        return (location_addr, location_pcc, postal_addr, postal_pcc)
+
+    if m := p4.match(address):
+        postal_addr = m.group(1) + '\n' + m.group(2)
+        postal_pcc = m.group(3)
+        return (location_addr, location_pcc, postal_addr, postal_pcc)
+
+    if m := p1.match(address):
+        postal_addr = m.group(1)
+        postal_pcc = m.group(2)
+        return (location_addr, location_pcc, postal_addr, postal_pcc)
+
+    if m := p5.match(address):
+        postal_addr = m.group(1)
+        if m.group(2):
+            postal_addr += f'{m.group(2)}'
+        return (location_addr, location_pcc, postal_addr, postal_pcc)
+
+    # default
+    print(f'*** parse_and_split_address_field: no match found for address:'
+          f' {address}')
+    return location_addr, location_pcc, postal_addr, postal_pcc
+
+
+@upgrade_task('ogc-966 extend agency and person tables with more fields')
+def extend_agency_and_person_with_more_fields(context):
+    ################################################################
+    # add columns to table 'agencies'
+    agencies_columns = ['email', 'phone', 'phone_direct', 'website',
+                        'location_address', 'location_code_city',
+                        'postal_address', 'postal_code_city',
+                        'opening_hours']
+    table = 'agencies'
+
+    for column in agencies_columns:
+        print(f'Checking table {table} for column {column} .. ')
+        if not context.has_column(table, column):
+            print(' -> column added')
+            context.add_column_with_defaults(
+                'agencies',
+                Column(column, Text, nullable=True),
+                default=lambda x: ''
+            )
+        else:
+            print(' -> column exists')
+
+    print()
+    context.session.flush()
+
+    ################################################################
+    # add columns to table 'people'
+    people_columns = ['location_address', 'location_code_city',
+                      'postal_address', 'postal_code_city', 'website_2']
+    table = 'people'
+
+    for column in people_columns:
+        print(f'Checking table {table} for column {column} .. ')
+        if not context.has_column(table, column):
+            # print(f'Adding column {column} to table {table}..')
+            print(' -> column added')
+            context.add_column_with_defaults(
+                'people',
+                Column(column, Text, nullable=True),
+                default=lambda x: ''
+            )
+        else:
+            print(' -> column exists')
+
+    print(f"Migrate data from table {table} column 'address' field to "
+          f"'location_address', 'location_code_city', 'postal_address' and "
+          f"'postal_code_city ..")
+    for person in context.session.query(Person):
+        if not person.address:
```

Comment:
This has the potential to blow up during upgrade, which is the worst moment because then you have to manually fix and complete the upgrade (or somebody else). In general, you should avoid using models in upgrade steps, since they might have been changed in the meantime (some of our servers have updates only every half year). In the future, we should add some functionality for reliable data migration (e.g. using historical models), I think there is a ticket somewhere. In the meantime, I use this (error prone) pattern for data migration:

1. Add schema migration in an upgrade step + a CLI command for data migration.
2. Upgrade the server.
3. Run the CLI command (on the servers they need them)
4. Add a post-migration schema migration (e.g. dropping now obsolete columns)

Reply:
@msom Can you please point me to such an example to get an idea how it can be done (the error prone way)?

Reply:
Have a look here: https://github.com/OneGov/onegov-cloud/commit/cda513c9e713d36d894037a76cb667ea35dd0025 (only CLI relevant; there is no schema migration because newly added tables are created automatically) + Post-Step: https://github.com/OneGov/onegov-cloud/commit/73573b2e0fbeab7647ffbb96902d4397f4bb9421

Reply:
Understood that I should separate things. One open item is how to make sure the CLI command will be executed during next roll-out (and only once)?
And if I understood right, I would need a separate branch to do the 'post' (cleanup) steps for a further release.

Reply:
By running it after the upgrade (on those server that need that) by hand. Yes, the post step would be added to the next release.

We could think about to adding a similar framework for data migration. In comparison, this should be (1) much more robust and never break upgrades and (2) allow migrations for models, even when they have changed.

Reply:
Hope, my last changes are a big step for the separation of concern :smile: